### PR TITLE
initial for eth0 support

### PR DESCRIPTION
**Pull request by sistemicorp**
Working prototype to support `eth0` as a connection option.

### CODE REVIEW

File: comitup/nmmon.py [hunk 1]
```diff
@@ -78,7 +80,29 @@ def cb_to(cb):
...
+    global wired_is_connected
+
+    log.info("MGAG: wired_changed_state {}".format(state))
+    from comitup.states import set_state
+
+    if state in PASS_STATES:
+        log.info("MGAG nmm - primary pass, set_state CONNECTED")
+        set_state('CONNECTED')
+        wired_is_connected = True
+        #send_cb(nm_dev_connect)
```
Comment (sistemicorp) [inline comitup/nmmon.py:93]: This code is not "right".
Should not be calling `set_state()` directly, but should be going thru `send_cb()`.
However I wasn't able to wrap my head around doing that... I will need to try again.

**Comment by sistemicorp** [comment]
This is "working" for me now.  That is, `eth0` is now a connection option. The external callback will get "CONNECTED" when wired is connected.  Plug'n'Play between `eth0` and `wlan0` seems to be working.

The `threading.Timer` is a hack to resolve the `eth0` state at boot up.

I spent a few hours today trying to make this work following the `send_cb(nm_dev_connect)` convention.  But I failed again.

**Comment by davesteele** [comment]
A quick looks suggests this could work OK.
Before merging, I would need to see a way for this to be optional via a config variable.

File: conf/comitup.conf [hunk 2]
```diff
@@ -69,3 +69,13 @@
 # The primary device is used to spawn the access point.
 #
 # primary_wifi_device: wlan0
+
+# manage_wired_device
+#
+# By default, eth0 (wired) device is not managed by comitup, it is ignored.
+# By setting manage_wired_device, comitup will also monitor eth0 connection state.
```
Comment (davesteele) [inline conf/comitup.conf:76]: 1) Make this "By _defining_ manage_wired_device ..."
2) "manage" is the wrong term. This change doesn't cause eth0 to be managed any differently. "monitor_wired_device"?

**Review by davesteele — CHANGES_REQUESTED** [review]

File: comitup/nm.py [hunk 3]
```diff
@@ -89,8 +90,11 @@ def device_name(device):
 
 def get_wired_device():
     dev = [x for x in get_devices() if x.DeviceType == 1]
-    if dev: return dev[0]
-    return None
+
+    if not dev or not getattr(conf, 'manage_wired_device', False):
```
Comment (davesteele) [inline comitup/nm.py:94]: I'd like to see the key initialized in the original config invocation, set to false

Comment (davesteele) [inline comitup/nm.py:94]: I'd like to see the key initialized in the original config invocation, set to false

Comment (sistemicorp) [reply comitup/nm.py:94]: Yes, its done that way - I shouldnot have pushed so early, but I switched machines and it was easiest way to move the code over.  Soon I have a better push.

File: conf/comitup.conf [hunk 4]
```diff
@@ -69,3 +69,13 @@
 # The primary device is used to spawn the access point.
 #
 # primary_wifi_device: wlan0
+
+# monitor_wired_device
+#
+# By default, eth0 (wired) device is not managed by comitup, it is ignored.
```
Comment (sistemicorp) [inline conf/comitup.conf:75]: change `not managed` to `not monitored`.

**Comment by sistemicorp** [comment]
On boot up, I am seeing inconsistent result that the wired CONNECTED state is not being posted to the external callback.

The first log from case of not working.  My callback log lines are `cmd: ...` (which I manually inserted here for timing),

        Starting Tue Apr  6 13:18:28 EDT 2021
        2021-04-06 13:18:40,472 - comitup - INFO - Starting comitup
        2021-04-06 13:18:40,633 - comitup - INFO - using SSID: p1125-man-asy
        2021-04-06 13:18:40,848 - comitup - INFO - Setting state to HOTSPOT
        2021-04-06 13:18:40,864 - comitup - INFO - Activating hotspot
        cmd: HOTSPOT Tue Apr  6 13:18:41 EDT 2021 ----------------------------------
        2021-04-06 13:18:45,741 - comitup - INFO - nmm - wired CONNECTED
        2021-04-06 13:18:47,263 - comitup - INFO - Setting state to CONNECTED
        // CONNECTED state never gets posted

This is a second boot, and working,

        2021-04-06 13:27:43,171 - comitup - INFO - Starting comitup
        2021-04-06 13:27:43,321 - comitup - INFO - using SSID: p1125-man-asy
        2021-04-06 13:27:43,547 - comitup - INFO - Setting state to HOTSPOT
        2021-04-06 13:27:43,565 - comitup - INFO - Activating hotspot
        cmd: HOTSPOT Tue Apr  6 13:27:44 EDT 2021 ----------------------------------
        2021-04-06 13:27:48,426 - comitup - INFO - nmm - wired CONNECTED
        2021-04-06 13:27:49,864 - comitup - INFO - Setting state to CONNECTED
        cmd: CONNECTED Tue Apr  6 13:27:50 EDT 2021 ----------------------------------

Because of the log line `INFO - nmm - wired CONNECTED` I believe `set_state('CONNECTED')` is being called,

    if state in [NetworkManager.NM_DEVICE_STATE_ACTIVATED, ]:
        wired_is_connected = True
        log.info("nmm - wired CONNECTED")
        set_state('CONNECTED')
    elif wired_is_connected:
        wired_is_connected = False
        set_state('HOTSPOT')
        log.debug("nmm - wired {}".format(state))
    else:
        log.debug("nmm - wired state {}".format(state))

**Review by davesteele — APPROVED** [review]

**Comment by sistemicorp** [comment]
This last change seemed to have fixed `eth0` CONNECTED state transition.  Before this change, many times the `eth0` state change was not propagated to the external callback, although the state was queued.  I think because the wifi device was "enabled", its state trumped the `eth0` state.

Q: Does the [4fb5c95](https://github.com/davesteele/comitup/pull/159/commits/4fb5c9563fb07786afd85b5e1a989f2727ab79f9) change make sense?  (apologise upfront for not be able to follow the callback/queuing system)

I will perform more tests...

**Comment by davesteele** [comment]
> Q: Does the [4fb5c95](https://github.com/davesteele/comitup/pull/159/commits/4fb5c9563fb07786afd85b5e1a989f2727ab79f9) change make sense?

I don't think so. It can fail if the Ethernet changes state after the enable call.

Maybe if the check were added as a wrapper around the \_fn's, or the functions themselves did the wired check.

**Comment by sistemicorp** [comment]
If Ethernet changes state after the enable call, doesn't `wired_changed_state()` get called and then `set_state('HOTSPOT')` would be called and then the "normal" AP system is back in place...

    def wired_changed_state(state, *args):
        global wired_is_connected
    
        from comitup.states import set_state
    
        if state in [NetworkManager.NM_DEVICE_STATE_ACTIVATED, ]:
            wired_is_connected = True
            log.info("nmm - wired CONNECTED")
            set_state('CONNECTED')
        elif wired_is_connected:
            wired_is_connected = False
            set_state('HOTSPOT')

**Comment by davesteele** [comment]
It's the race conditions that are mean.

**Comment by davesteele** [comment]
I haven't looked at your latest in detail, but I have taken a look at how I might go about it:
  * Add a comitup.conf variable to let NetworkManager manage the CONNECTED state, with an explanation of what that means. Default to unset, which is interpreted as False.
  * in nmmon.py set_device_listeners, add a listener with callback nm_changed_state() looking at org.freedesktop.NetworkManager [StateChanged()](https://developer.gnome.org/NetworkManager/stable/gdbus-org.freedesktop.NetworkManager.html#gdbus-signal-org-freedesktop-NetworkManager.StateChanged). Add the callback as an argument to enable(). Note that disable() is unused.
  * Define the callback in states.py. Use the callback, and an initial seed, to track [State](https://developer.gnome.org/NetworkManager/stable/gdbus-org.freedesktop.NetworkManager.html#gdbus-property-org-freedesktop-NetworkManager.State).
  * Make a utility fn in states.py, nm_connected(), which returns True if State == [NM_STATE_CONNECTED_GLOBAL](https://developer.gnome.org/NetworkManager/stable/nm-dbus-types.html#NMState) and if comitup.conf indicates that we want NM to manage the CONNECTED state.
  * The callback mentioned earlier, if State == NM_STATE_CONNECTED_GLOBAL and nm_connected() == True and state == HOTSPOT, sets the state to CONNECTING.
  * connecting_start() immediately sets state to CONNECTED if nm_connected() is True.
  * connected_timeout() (and maybe connected_fail()) will not transition out of CONNECTED if nm_connected() == True

**Comment by sistemicorp** [comment]
Thank you Dave.  Those hints may help me.  I spent the last few days trying to get things to work, and I tried using more of the states and architecture that WiFi was using.... but I failed... I have a commit of it, full of logging, but I didn't push it.

Today I backtracked and trying your original suggestion of managing `eth0` externally and start/stop `comitup` service.  I think I am making good headway on that... know in a few hours or tomorrow.

This pull request should probably be closed.

**Comment by davesteele** [comment]
I appreciate the effort.

**Pull request closed without merging**